src/Orleans/Lifecycle/ILifecycleObservable.cs

```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+namespace Orleans
+{
+    /// <summary>
+    /// Observable lifecycle.
+    /// Each stage of lifecycle is observable.  All observers will be notified 
+    ///   when the stage is reached when starting, and stopping.
+    /// Stages are started in ascending order, and stopped in decending order.
+    /// </summary>
+    /// <typeparam name="TStage"></typeparam>
+    public interface ILifecycleObservable<in TStage>
+    {
+        /// <summary>
+        /// Subscribe for notification when a stage is reached while starting or stopping.
+        /// </summary>
+        /// <param name="stage">stage of interest</param>
+        /// <param name="observer">stage observer</param>
+        /// <returns>A disposable that can be disposed of to unsubscribe</returns>
+        IDisposable Subscribe(TStage stage, ILifecycleObserver observer);
+    }
```

Comment:
I'm open to other approaches, but this pattern will be very familiar and comfortable to those familiar with IObservable which this is somewhat modeled on.
https://msdn.microsoft.com/en-us/library/dd990377(v=vs.110).aspx

The general observer pattern usually involves an attach and detach, which is a viable approach as well, but a bit more cumbersome to use.  We can always use extension functions to make it more usable though, so I'd not dismiss this out of hand, should anyone want to make an argument for it.
http://www.dofactory.com/net/observer-design-pattern

We use subscription handles in streams primarily because the unsubscribe operation is asynchronous so a disposable would not work, and since we had to deviate from the IObservable conventions, we added more stuff to the handle, but I'm not yet convinced that complex handles are a good idea, in general.